Fix Delete from product_batch on product delete

fappels opened the pull request:
Data from product_batch table needs to be deleted before product is deleted from product_stock because of fk_product_batch_fk_product_stock constraint.

fappels committed:
Delete from product_batch on product delete

defrance commented:
NO 
we need to add a control in isObjectUsed function (add product_batch in $childtables array)
and made manually suppress of product_batch before (like for the stock ???)...

fappels commented:
product_batch has no fk_product (is extension of product_stock). Updating all product_batch data with fk_product on migration a not a good idea.

defrance commented:
I explain my point of view : 

First your delete must be made just on adding 'product_batch' on elements array just near

Second : before delete product we need to control all stock elements is empty with this product and they are no movement based on this product. 
Suppress product with stock and movements made must be not allowed if product_stock and stock_movement contain line associated to this product and must controlled before allow suppression

please create isProductInStock and isProductMovement functions and use it to prevent this

defrance commented:
please consider this issues : https://github.com/Dolibarr/dolibarr/issues/5998

eldy commented on htdocs/product/class/product.class.php:
```diff
@@ -924,7 +924,23 @@ function delete($id=0, $notrigger=0)
-
+			
+			// Delete from product_batch on product delete	
+			if (! $error && $this->hasbatch())
```
The test $this->hasbatch() should be removed.
If flag "need serial/lot number" is on, it will be true
And if flag "need serial/lot number", we must also do the delete even if there is a high probability that table is empty.
Because flag may have been modified after beeing set. 
Don't you think so ?

fappels replied:
Done, you are correct. Intention is to remove all data linked to product_stock, which will be
deleted.

fappels committed:
Remove hasbatch test

Intention is to remove all data linked to product_stock, which will be
deleted.

fappels commented:
Also closes #4565
